cluster mempool: introduce TxGraph

sipa opened the pull request:
Part of cluster mempool: #30289.

### 1. Overview

This introduces the `TxGraph` class, which encapsulates knowledge about the (effective) fees, sizes, and dependencies between all mempool transactions, but nothing else. In particular, it lacks knowledge about `CTransaction`, inputs, outputs, txids, wtxids, prioritization, validatity, policy rules, and a lot more. Being restricted to just those aspects of the mempool makes the behavior very easy to fully specify (ignoring the actual linearizations produced), and write simulation-based tests for (which are included in this PR).

### 2. Interface

The interface can be largely categorized into:
* Mutation functions:
  * `AddTransaction` (add a new transaction with specified feerate, and get a `Ref` object back to identify it).
  * `RemoveTransaction` (given a `Ref` object, remove the transaction).
  * `AddDependency` (given two `Ref` objects, add a dependency between them).
  * `SetTransactionFee` (modify the fee associated with a Ref object).
* Inspector functions:
  * `GetAncestors` (get the ancestor set in the form of `Ref*` pointers)
  * `GetAncestorsUnion` (like above, but for the union of ancestors of multiple `Ref*` pointers)
  * `GetDescendants` (get the descendant set in the form of `Ref*` pointers)
  * `GetDescendantsUnion` (like above, but for the union of ancestors of multiple `Ref*` pointers)
  * `GetCluster` (get the connected component set in the form of `Ref*` pointers, in the order they would be mined).
  * `GetIndividualFeerate` (get the feerate of a transaction)
  * `GetChunkFeerate` (get the mining score of a transaction)
  * `CountDistinctClusters` (count the number of distinct clusters a list of `Ref`s belong to)
* Staging functions:
  * `StartStaging` (make all future mutations operate on a proposed transaction graph)
  * `CommitStaging` (apply all the changes that are staged)
  * `AbortStaging` (discard all the changes that are staged)
* Miscellaneous functions:
  * `DoWork` (do queued-up computations now, so that future operations are fast)

This `TxGraph::Ref` type used as a "handle" on transactions in the graph can be inherited from, and the idea is that in the full cluster mempool implementation (#28676, after it is rebased on this), `CTxMempoolEntry` will inherit from it, and all actually used Ref objects will be `CTxMempoolEntry`s. With that, the mempool code can just cast any `Ref*` returned by txgraph to `CTxMempoolEntry*`.

### 3. Implementation

Internally the graph data is kept in clustered form (partitioned into connected components), for which linearizations are maintained and updated as needed using the `cluster_linearize.h` algorithms under the hood, but this is hidden from the users of this class. Implementation-wise, mutations are generally applied lazily, appending to queues of to-be-removed transactions and to-be-added dependencies, so they can be batched for higher performance. Inspectors will generally only evaluate as much as is needed to answer queries, with roughly 5 levels of processing to go to fully instantiated and acceptable cluster linearizations, in order:
1. `ApplyRemovals` (take batches of to-be-removed transactions and translate them to "holes" in the corresponding Clusters/DepGraphs).
2. `SplitAll` (creating holes in Clusters may cause them to break apart into smaller connected components, so make turn them into separate Clusters/linearizations).
3. `GroupClusters` (figure out which Clusters will need to be combined in order to add requested to-be-added dependencies, as these may span clusters).
4. `ApplyDependencies` (actually merge Clusters as precomputed by `GroupClusters`, and add the dependencies between them).
5. `MakeAcceptable` (perform the LIMO linearization algorithm on Clusters to make sure their linearizations are acceptable).

### 4. Future work

This is only an initial version of TxGraph, and some functionality is missing before #28676 can be rebased on top of it:
* The ability to get comparative feerate diagrams before/after for the set of staged changes (to evaluate RBF incentive-compatibility).
* Mining interface (ability to iterate transactions quickly in mining score order) (see #31444).
* Eviction interface (reverse of mining order, plus memory usage accounting) (see #31444).
* Ability to fix oversizedness of clusters (before or after committing) - this is needed for reorgs where aborting/rejecting the change just is not an option (see #31553).
* Interface for controlling how much effort is spent on LIMO. In this PR it is hardcoded.

Then there are further improvements possible which would not block other work:
* Making Cluster a virtual class with different implementations based on transaction count (which could dramatically reduce memory usage, as most Clusters are just a single transaction, for which the current implementation is overkill).
* The ability to have background thread(s) for improving cluster linearizations.

sipa committed:
txgraph: Add initial version (feature)

This adds an initial version of the txgraph module, with the TxGraph class.
It encapsulates knowledge about the fees, sizes, and dependencies between all
mempool transactions, but nothing else.

In particular, it lacks knowledge about txids, inputs, outputs, CTransactions,
... and so forth. Instead, it exposes a generic TxGraph::Ref type to reference
nodes in the TxGraph, which can be passed around and stored by layers on top.

sipa committed:
clusterlin: Add FixLinearization function + fuzz test

This function takes an existing ordering for transactions in a DepGraph, and
makes it a valid linearization for it (i.e., topological). Any topological
prefix of the input remains untouched.

sipa committed:
clusterlin: Make IsAcyclic() a DepGraph member function

... instead of being a separate test-only function.

Also add a fuzz test for it returning false.

sipa committed:
txgraph: Expose ability to compare transactions (feature)

In order to make it possible for higher layers to compare transaction quality
(ordering within the implicit total ordering on the mempool), expose a comparison
function and test it.

sipa committed:
txgraph: Special-case removal of tail of cluster (Optimization)

When transactions are removed from the tail of a cluster, we know the existing
linearization remains acceptable (if it already was), but may just need splitting
and postlinearization, so special case these into separate quality levels.

sipa committed:
txgraph: Add internal sanity check function (tests)

To make testing more powerful, expose a function to perform an internal sanity
check on the state of a TxGraph. This is especially important as TxGraphImpl
contains many redundantly represented pieces of information:

* graph contains clusters, which refer to entries, but the entries refer back
* graph maintains pointers to Ref objects, which point back to the graph.

This lets us make sure they are always in sync.

sipa committed:
txgraph: Delay chunking while sub-acceptable (optimization)

Chunk-based information (primarily, chunk feerates) are never accessed without
first bringing the relevant Clusters to an "acceptable" quality level. Thus,
while operations are ongoing and Clusters are not acceptable, we can omit
computing the chunkings and chunk feerates for Clusters.

sipa committed:
txgraph: Add simulation fuzz test (tests)

This adds a simulation fuzz test for txgraph, by comparing with a naive
reimplementation that models the entire graph as a single DepGraph, and
clusters in TxGraph as connected components within that DepGraph.

sipa committed:
txgraph: Group per-graph data in ClusterSet (refactor)

This is a preparation for a next commit where a TxGraph will start representing
potentially two distinct graphs (a main one, and a staging one with proposed
changes).

sipa committed:
txgraph: Make max cluster count configurable and "oversize" state (feature)

Instead of leaving the responsibility on higher layers to guarantee that
no connected component within TxGraph (a barely exposed concept, except through
GetCluster()) exceeds the cluster count limit, move this responsibility to
TxGraph itself:
* TxGraph retains a cluster count limit, but it becomes configurable at construction
  time (this primarily helps with testing that it is properly enforced).
* It is always allowed to perform mutators on TxGraph, even if they would cause the
  cluster count limit to be exceeded. Instead, TxGraph exposes an IsOversized()
  function, which queries whether it is in a special "oversize" state.
* During oversize state, many inspectors are unavailable, but mutators remain valid,
  so the higher layer can "fix" the oversize state before continuing.

sipa committed:
txgraph: Abstract out ClearLocator (refactor)

Move a number of related modifications to TxGraphImpl into a separate
function for removal of transactions. This is preparation for a later
commit where this will be useful in more than one place.

sipa committed:
txgraph: Destroying Ref means removing transaction (feature)

Before this commit, if a TxGraph::Ref object is destroyed, it becomes impossible
to refer to, but the actual corresponding transaction node in the TxGraph remains,
and remains indefinitely as there is no way to remove it.

Fix this by making the destruction of TxGraph::Ref trigger immediate removal of
the corresponding transaction in TxGraph, both in main and staging if it exists.

sipa committed:
txgraph: Add staging support (feature)

In order to make it easy to evaluate proposed changes to a TxGraph, introduce a
"staging" mode, where mutators (AddTransaction, AddDependency, RemoveTransaction)
do not modify the actual graph, but just a staging version of it. That staging
graph can then be commited (replacing the main one with it), or aborted (discarding
the staging).

jonatack commented:
Concept ACK

ismaelsadeeq reviewed (commented):
Concept ACK


> GetChunkFeerate (get the mining score of a transaction)


From reading the description and skimming through the code, IIUC `TxGraph` encapsulates knowledge about fees and sizes but lacks knowledge about prioritization. This implies that the fee it uses is the effective fee, not the modified fee that includes prioritization.

How are we accounting for the real mining score of a  `Ref` without that prioritization knowledge? The current mining algorithm uses the modified ancestor fee to calculate the mining score.

Could you clarify how prioritization will be handled post-cluster mempool?

ismaelsadeeq commented:
Thank you for clarifying. Reading 'lacks knowledge about prioritization' led me to infer that.
But I see now you are talking about `mapDeltas`

sipa committed:
txgraph: Add DoWork function (feature)

This can be called when the caller has time to spend now, and wants future operations
to be fast.

sipa committed:
txgraph: Add CountDistinctClusters function (feature)

sipa committed:
txgraph: Avoid looking up the same child cluster repeatedly (optimization)

Since m_deps_to_add has been sorted by child Cluster* already, all dependencies
with the same child will be processed consecutively. Take advantage of this by
remember the last partition merged with, and reusing that if applicable.

sipa committed:
txgraph: Avoid representative lookup for each dependency (optimization)

The m_deps_to_add vector is sorted by child Cluster*, which matches the
order of an_clusters. This means we can walk through m_deps_to_add while
doing the representative lookups for an_clusters, and reuse them.

instagibbs commented:
> Unlinking only happens when the calling code destroys a Ref. Currently, unlinking can happen either through Ref destruction, or through TxGraph::Cleanup() (which unlinks anything not appearing in the graph anymore).

this would result in:
1) exists
1) removed
2) unlinked + destroyed

as the three possible states?

sipa committed:
txgraph: Cache oversizedness of graphs (optimization)

sipa committed:
txgraph: Avoid per-group vectors for clusters & dependencies (optimization)

Instead construct a single vector with the list of all clusters in all groups,
and then store per-group offset/range in that list.

For dependencies, reuse m_deps_to_add, and store offset/range into that.

ismaelsadeeq commented on src/txgraph.h:
```diff
@@ -12,8 +12,7 @@
 
-/** No connected component within TxGraph is allowed to exceed this number of transactions. */
-static constexpr unsigned CLUSTER_COUNT_LIMIT{64};
+static constexpr unsigned MAX_CLUSTER_COUNT_LIMIT{64};
```
In  "txgraph: (feature) make max cluster count configurable and "oversize" state" 15da6fdd19813a08bf9571147f97e22b5390ecf9

Maybe comment why 64 selected and how having limit above this will hurt the node/miner? fwiw we have seen clusters with size more than 64 in the past https://bitcoincore.reviews/25038, but I think it's not a regular occurrence.

sipa committed:
feefrac: Introduce tagged wrappers to distinguish vsize/WU rates

sipa committed:
scripted-diff: (refactor) ClusterIndex -> DepGraphIndex

Since cluster_linearize.h does not actually have a Cluster type anymore, it is more
appropriate to rename the index type to DepGraphIndex.

-BEGIN VERIFY SCRIPT-
sed -i 's/Data type to represent transaction indices in clusters./Data type to represent transaction indices in DepGraphs and the clusters they represent./' $(git grep -l 'using ClusterIndex')
sed -i 's|\<ClusterIndex\>|DepGraphIndex|g' $(git grep -l 'ClusterIndex')
-END VERIFY SCRIPT-

sipa committed:
txgraph: Multiple inputs to Get{Ancestors,Descendant}Refs (preparation)

This is a preparation for the next commit, which adds a feature to request
the Refs to multiple ancestors/descendants at once.

ismaelsadeeq reviewed (commented):
Code review to 741a6a8c4d851cb10ecee810a09187bcbfa5af4c

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -1118,3 +1118,58 @@ FUZZ_TARGET(clusterlin_merge)
+            reader >> VARINT(val);
+        } catch (const std::ios_base::failure&) {}
+        val %= todo.Count();
+        // Find which element in todo that corresponds to.
```
Yes this is much better now!
Thanks

instagibbs reviewed (commented):
reviewed up to "txgraph: (feature) add initial version" 781c15bfca1ebaffe7b634196e19144f5ab10a50 , basically all nits at this point, logic all makes sense

One question I had coming up is use of `Assume()`, there is a number of places that `Assume()` is used, then subsequently may result in UB or weird behavior. In a few spots I gave suggestions but was wondering if there is thinking behind doing more (or Assert'ing) or not to handle that in release builds.

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -1110,6 +1121,106 @@ void TxGraphImpl::SetTransactionFee(const Ref& ref, int64_t fee) noexcept
+        assert(entry.m_locator.index == lin_pos);
+        // Check linearization position and chunk feerate.
+        if (!linchunking.GetChunk(0).transactions[lin_pos]) {
+            linchunking.MarkDone(linchunking.GetChunk(0).transactions);
```
In  "txgraph: (tests) add internal sanity check function" 741a6a8c4d851cb10ecee810a09187bcbfa5af4c

Should we also compare that the list of transactions in chunk 0 matches `m_linearization` subset from from previous hit of the conditional statement to `lin_pos`?

ismaelsadeeq replied:
Yeah I was thinking something like

<details>
<summary>diff</summary>

```diff
diff --git a/src/txgraph.cpp b/src/txgraph.cpp
index afaa46e6326..a9b3e446d26 100644
--- a/src/txgraph.cpp
+++ b/src/txgraph.cpp
@@ -1143,6 +1143,7 @@ void Cluster::SanityCheck(const TxGraphImpl& graph) const
 
     // Verify m_linearization.
     SetType m_done;
+    SetType last_chunk;
     assert(m_depgraph.IsAcyclic());
     for (auto lin_pos : m_linearization) {
         assert(lin_pos < m_mapping.size());
@@ -1155,9 +1156,12 @@ void Cluster::SanityCheck(const TxGraphImpl& graph) const
         assert(entry.m_locator.index == lin_pos);
         // Check linearization position and chunk feerate.
         if (!linchunking.GetChunk(0).transactions[lin_pos]) {
+            assert(linchunking.GetChunk(0).transactions == last_chunk);
+            last_chunk = SetType();
             linchunking.MarkDone(linchunking.GetChunk(0).transactions);
         }
         assert(entry.m_chunk_feerate == linchunking.GetChunk(0).feerate);
+        last_chunk.Set(lin_pos);
         // If this Cluster has an acceptable quality level, its chunks must be connected.
         if (m_quality == QualityLevel::ACCEPTABLE || m_quality == QualityLevel::OPTIMAL) {
             assert(m_depgraph.IsConnected(linchunking.GetChunk(0).transactions));
```
</details>

ismaelsadeeq commented on src/test/fuzz/txgraph.cpp:
```diff
@@ -0,0 +1,430 @@
+        /** The number of possible choices. */
+        size_t choices = tx_count + sim.removed.size() + 1;
+        /** Pick one of them. */
+        auto choice = provider.ConsumeIntegralInRange<size_t>(0, choices - 1);
```
Makes sense thanks.

ismaelsadeeq commented on src/test/fuzz/txgraph.cpp:
```diff
@@ -0,0 +1,430 @@
+    auto real = MakeTxGraph();
+    SimTxGraph sim;
+
+    /** Function to pick any Ref (in sim real, sim.removed, or empty). */
```
nit: There is still one remaining in line 180.

ismaelsadeeq commented on src/test/fuzz/txgraph.cpp:
```diff
@@ -0,0 +1,430 @@
+        return &empty_ref;
+    };
+
+    LIMITED_WHILE(provider.remaining_bytes() > 0, 200) {
```
Yeah something like that that 

In most of the fuzz test in `src/test/fuzz/cluster_linearize.cpp` you provided a really nice description , which is just an overview of what the test is going to do before jumping to the instructions.

Here we just jump into the instructions without that context.

But this is nitty, you can ignore this comment and resolve.

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -0,0 +1,1150 @@
+            // The existing Cluster is an entire component. Leave it be, but update its quality.
+            graph.SetClusterQuality(m_quality, m_setindex, QualityLevel::NEEDS_RELINEARIZE);
+            // We need to recompute and cache its chunking.
+            Updated(graph);
```
I see this now 👍🏾

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -311,40 +373,50 @@ class TxGraphImpl final : public TxGraph
-void TxGraphImpl::ClearLocator(GraphIndex idx) noexcept
+void TxGraphImpl::ClearLocator(int level, GraphIndex idx) noexcept
 {
     auto& entry = m_entries[idx];
```
```suggestion
    Assume (level - MAX_LEVELS - 1);
    auto& entry = m_entries[idx];
```

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -356,14 +428,65 @@ void Cluster::Updated(TxGraphImpl& graph) noexcept
+        // For every transaction Entry in this Cluster, if it also exists in a lower-level Cluster,
+        // then that Cluster conflicts.
+        if (entry.m_locator[m_level - 1].IsPresent()) {
+            out.push_back(entry.m_locator[m_level - 1].cluster);
```
Should be set to avoid duplicates, will prevent de duplicating later on?

ismaelsadeeq replied:
Agreed, I calculated recently.
This is better.

The constant factors for GetDistinctClusters implementation using vector and then sorting is also better 👍🏾
Please resolve.

ismaelsadeeq reviewed (commented):
Code Review e19bbc328236f64716034277857951184309cd14

It seems to me that this is designed with the flexibility to accommodate more levels, which explains the use of `std::vector` for `ClusterSet`

If yes then current approach is perfect else would it be better for `TxGraph` to maintain a `std::array` of two `ClusterSet? `main` and `staging and store their indexes in `TxGraph` as enums? This way, `Cluster` would only need to maintain that enum.  

In `ClusterSet`, we could introduce a boolean to indicate whether a set is active or not.  

The `main` set could always remain active, while `staging` would be flexible and toggleable.  

Thus, when we start staging, we can simply mark the back array as active, and once done, clear the back array and mark it as inactive.  

I believe this approach would reduce Assumes since we explicitly know we can only have two ClusterSet's.

Pumpkin1234567812 commented:
ApplyDependencies(0);

instagibbs reviewed (commented):
Looking good 11135464c5c3aef1ce8d2823120467a522ce2c87

I just have conceptual issues with what PostLinerize is guaranteeing us beyond what is explicitly stated in the header file. 

Pile of nits/questions otherwise

sipa reviewed (commented):
Addressed most comments. I will address the `PostLinearize` related ones later.

instagibbs commented on src/txgraph.h:
```diff
@@ -143,10 +143,16 @@ class TxGraph
      *  staging graph exists, it is queried; otherwise the main graph is queried. The queried
      *  graph must not be oversized. Returns {} if arg does not exist in the queried graph. */
     virtual std::vector<Ref*> GetDescendants(const Ref& arg, bool main_only = false) noexcept = 0;
+    /** Like GetDescendants, but return the Refs for all transactions in the union of the provided
```
major reading comprehension issues it seems

instagibbs commented on src/txgraph.cpp:
```diff
@@ -189,28 +209,59 @@ class TxGraphImpl final : public TxGraph
         std::vector<std::pair<GraphIndex, GraphIndex>> m_deps_to_add;
         /** Information about the merges to be performed, if known. */
         std::optional<GroupData> m_group_data = GroupData{};
+        /** Which entries were removed in this ClusterSet (so they can be wiped on abort). */
```
yes, thanks

instagibbs commented on src/txgraph.cpp:
```diff
@@ -603,56 +755,102 @@ std::unique_ptr<Cluster> TxGraphImpl::ExtractCluster(QualityLevel quality, Clust
+    // Copy the Cluster from the level it was found at to higher levels, if any.
+    while (level < to_level) {
+        // Make the Cluster Acceptable before copying. This isn't strictly necessary, but doing it
+        // now avoids doing doable work later.
```
hah, was my first guess that I doubted :)

instagibbs commented on src/txgraph.cpp:
```diff
@@ -1176,101 +1400,201 @@ FeePerWeight Cluster::GetIndividualFeerate(DepGraphIndex idx) noexcept
+    // Mark are removed transactions as Missing (so the stage_level locator for these transactions
+    // can be reused if another staging is created).
+    for (auto idx : stage.m_removed) {
+        m_entries[idx].m_locator[stage_level].SetMissing();
```
saw the updated comment/code in SanityCheck, makes sense

instagibbs commented:
verified changes, just waiting on further PostLinearization commentary

via `git range-diff master 11135464c5c3aef1ce8d2823120467a522ce2c87 d5abb86439e79d4adfbfbd46f833268bbca0bf6e`

yancyribbens commented on src/cluster_linearize.h:
```diff
@@ -1336,6 +1336,35 @@ std::vector<ClusterIndex> MergeLinearizations(const DepGraph<SetType>& depgraph,
+        // Find which position to place elem in (updating j), continuously moving the elements
+        // in between forward.
+        while (place_before.Any()) {
+            auto to_swap = linearization[len - 1 - (j - 1)];
```
The comment ` // j cannot be 0 here;` implies this should be `j != 0` instead of `j > 0`.  Since it's unsigned `j != 0` is also correct.

instagibbs commented on src/txgraph.cpp:
```diff
@@ -365,26 +369,56 @@ void Cluster::ApplyRemovals(TxGraphImpl& graph, std::span<GraphIndex>& to_remove
+    QualityLevel new_quality = m_quality == QualityLevel::NEEDS_SPLIT ? QualityLevel::NEEDS_RELINEARIZE :
+                               m_quality == QualityLevel::NEEDS_SPLIT_OPTIMAL ? QualityLevel::OPTIMAL :
+                               QualityLevel::ACCEPTABLE;
+    // If the cluster was NEEDS_SPLIT_OPTIMAL, and we're thus going to produce OPTIMAL clusters, we
```
Update from offline conversation: This was a confusion about what we're calling "optimal". The splits clusters are already optimal in the sense of the diagram check, but not minimal or fully connected. PostLinearization makes sure that all chunks are connected/minimal.

Perhaps a rewording of this section to make it abundantly clear?

instagibbs replied:
Also, as discussed offline, this seems like an easy scenario to write a fuzz test for as described above ^

instagibbs commented:
ACK 72a97c0a07ea6e5a95ab37c8d95e1ea02cff8e92

With the weaker claims in the code re:PostLinearization, it works for me, and obviates the need for an additional harness test.

`git range-diff master d5abb86439e79d4adfbfbd46f833268bbca0bf6e 72a97c0a07ea6e5a95ab37c8d95e1ea02cff8e92`

ismaelsadeeq reviewed (approved):
Code review ACK 72a97c0a07ea6e5a95ab37c8d95e1ea02cff8e92 🛸 

I've reviewed each commit and also looked at how the interface is being used by `CTxMemPool` in the big PR

ajtowns reviewed (commented):
This seems fairly difficult to review -- there's a lot of mostly implicit background in what TxGraph does and why it does it that way, as evidenced by both the description in this PR and in the [review club](https://bitcoincore.reviews/31363).

When I started typing this comment, I was wondering if maybe it would make sense to make the commits here into more of a gentle introduction to TxGraph, rather than one commit that does everything, then optimisations. But I don't really think that approach would survive subsequent modifications to the code very well. Likewise for just adding more comments to the code.

Alternatively, I wonder if it would be helpful both for new devs and debugging to expose the TxGraph API as a python module, possibly with additional debugging handles so that you can examine internal state? I could imagine an interactive python notebook that allows you to interact with a real TxGraph being a useful experimental tool for learning/writing mempool code and also being something that could perhaps be reasonably maintained. [Here's](https://github.com/ajtowns/bitcoin/commits/202503-txgraph-python/) a rough take at what that might look like. It's using boost-python, which seems reasonable, but doesn't seem super friendly to pure-virtual classes (hence the tedious wrapper class) ~~or move-only objects (hence AddTransaction being wrapped via a constructor)~~. (EDIT: add link for "here's"; EDIT2: converting the `vector<Ref*>` stuff to python made the move-only Ref stuff irrelevant; the wrapper class isn't quite so tedious any more either really)

Comments below are mostly a bunch of nits/observations while familiarising myself with the code, feel free to ignore.

ajtowns reviewed (commented):
Preliminary ACK 72a97c0a07ea6e5a95ab37c8d95e1ea02cff8e92

I think it's okay to merge this code:
 * as of this PR, it isn't actually used, so should not introduce bugs immediately
 * the fuzz testing is pretty good, so there should not be any catastrophic bugs even if it were used
 * while maybe the design could be improved, that can be done incrementally after it's merged
 * I'm reasonably confident I understand what's going on and can help find/fix bugs/problems if they occur
 * I think the design and implementation are a good step forward for implementing cluster mempool

On the other hand, calling this "preliminary" as:
 * I'm not yet confident that I could maintain this code if its authors all suddenly vanished
 * I'm a bit concerned that to contribute to mempool code you might need to be a triple expert -- C++, bitcoin, and industrial optimisation theory; not a show stopper, but as further PRs progress I'd like to gain confidence either that the optimisation parts can be safely treated as a very stable black box, or (ideally and) that we can reasonably onboard people to understand how the internals of the black box work.

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -0,0 +1,2076 @@
+    };
+
+    /** The ClusterSets in this TxGraphImpl. Has exactly 1 (main) or exactly 2 elements (main and staged). */
+    std::vector<ClusterSet> m_clustersets;
```
Curious to know this contrived use case, I also think it's a bit of an overkill in a previous comment https://github.com/bitcoin/bitcoin/pull/31363#pullrequestreview-2618143580, but @sipa made a compelling argument for it https://github.com/bitcoin/bitcoin/pull/31363#issuecomment-2666866811

ajtowns replied:
I think `StartStaging()` gets a bit simpler if you know that it can only be called with `m_clustersets.size()==1`:

```c++
 void TxGraphImpl::StartStaging() noexcept
 {
+    Assume(m_clustersets.size() == 1);  // vs < MAX_LEVELS
+    SplitAll(0); // vs m_clustersets.size() - 1
+    ApplyDependencies(0); // ditto
     m_clustersets.emplace_back();
+    auto& stage = m_clustersets[1]; // vs m_clustersets.back()
+    auto& main = m_clustersets[0]; // vs *(m_clustersets.rbegin() + 1);
     ...
```

That's the only use of `m_clustersets.rbegin()`, the remaining API of m_clustersets seems to be:

```c++
    class MainStageClusterSets
    {
    private:
        std::vector<ClusterSet> cs;

    public:
        MainStageClusterSets()
        {
            cs.reserve(MAX_LEVELS);
            cs.emplace_back();
        }

        bool UnfinishedTodo() const {
            for (auto& clusterset : cs) {
                if (!clusterset.m_deps_to_add.empty()) return true;
                if (!clusterset.m_to_remove.empty()) return true;
                if (!clusterset.m_removed.empty()) return true;
            }
            return false;
        }

        const ClusterSet& operator[](size_t i) const { return cs[i]; }
        ClusterSet& operator[](size_t i) { return cs[i]; }
        ClusterSet& back() { return cs.back(); }
        size_t size() const { return cs.size(); }

        void emplace_back() { cs.emplace_back(); }
        void pop_back() { cs.pop_back(); }
    };
    MainStageClusterSets m_clustersets;
```

which seem pretty straightforward to switch to a main/optional-stage model. Doing some renaming would make sense if so (emplace_back/pop_back to add/remove_staging; back to work maybe; size to has_staging maybe; maybe a templated ForEach() accepting a lambda taking a ClusterSet for iteration).

ajtowns commented:
> * However, it's also the case that if the compiler can prove a statement is side-effect free, it can optimize the call away in non-debug builds.

That makes sense. Might be a good idea to update doc/developer-notes.md pointing out that use case? It currently says:

> * `Assume` should be used to document assumptions when program execution can
>   safely continue even if the assumption is violated.

sipa committed:
txgraph: Add GetChunkFeerate function (feature)

This adds a function to query the chunk feerate of a transaction, by caching it
inside the Entry objects.

sipa reviewed (commented):
Big change: `std::vector<ClusterSet>` is gone, replaced by `m_main_clusterset` and `m_staging_clusterset`.
* New helper functions `GetTopLevel()`, `GetSpecifiedLevel()`, `GetClusterSet()` made this fairly easy to do.
* Simplifications resulted in `ClearLocator`, `GetConflicts`, `PullIn`, `MakeTransactionsMissing`, `StartStaging`, `CommitStaging`, `ExtractCluster`.
* `CopyTo` was renamed to `CopyToStaging`.
* `LevelDown` was renamed to `MoveToMain`.
* `MakeTransactionsMissing` was renamed to `MakeStagingTransactionsMissing`.

Other changes:
* Various comments
* `Ref` objects are now allowed to outlive `TxGraph`.
* `GetChunkFeerate` (and its testing, plus `Entry::m_chunk_feerate` maintenance) was moved to a separate commit.

ajtowns commented on src/txgraph.h:
```diff
@@ -0,0 +1,124 @@
+     *  further calls, only Refs created by AddTransaction() are allowed to be passed to this
+     *  TxGraph object (or empty Ref objects). */
+    [[nodiscard]] virtual Ref AddTransaction(const FeePerWeight& feerate) noexcept = 0;
+    /** Remove the specified transaction. This is a no-op if the transaction was already removed.
```
Maybe it would be helpful to explain in the header that AddTransaction() creates a Ref, which you're then expected to assign/move into your own container, which may be a subclass containing even more info about the tx, and then that serves as a permanent handle, even as the txgraph gets rearranged internally and that it's deleted from the graph either with RemoveTransaction explicitly or implicitly by destructing the Ref? Essentially expanding the "Data type used to reference transactions within a TxGraph" comment that's already there to be a bit more ELI5 maybe?

If you create an empty Ref (ie, construct it directly, not via AddTransaction) is there any way to associate that Ref with a TxGraph or is it forever useless as a Ref? Would it make any sense to have an AddTransaction variant that takes a Ref& argument rather than returning a new Ref? I guess equivalently, is an empty Ref useful for anything other than tests? Maybe "Non-empty Refs can only be created using AddTx" should be more explicit and say "You want to use AddTransaction() rather than constructing a Ref directly".

ajtowns commented on src/txgraph.cpp:
```diff
@@ -0,0 +1,2076 @@
+static constexpr int MAX_LEVELS{2};
+
+// Forward declare the TxGraph implementation class.
+class TxGraphImpl;
```
Sounds like this is something that can be reconsidered when we get a PR that introduces variant Clusters then. Mark as resolved for now?

ajtowns commented on src/txgraph.h:
```diff
@@ -0,0 +1,183 @@
+     *  for oversized graphs. The mutators above are always available. Removing a transaction by
+     *  destroying its Ref while staging exists will not clear main's oversizedness until staging
+     *  is aborted or committed. */
+    virtual bool IsOversized(bool main_only = false) noexcept = 0;
```
If I'm understanding right, this seem like it makes `IsOversized()` largely unfixable by the client -- so the only reasonable strategy I can see is to do:

```c++
    StartStaging();
    // change transactions
    if (IsOversized() || otherwise_undesirable()) {
        AbortStaging();
    } else {
        CommitStaging();`
    }
```

That's a very reasonable strategy of course, so this isn't a complaint! Just that if so, it seems like it should be documented a bit more clearly?  I wonder a bit if `CommitStaging()` should have an `Assume(G_FUZZING || !m_oversized)` or something...

(In particular, if you ever end up with the main graph being oversized, I don't see how you can even discover which txs are contributing to the oversized-ness in order to remove them -- `GetCluster(ref)` will fail the `m_deps_to_add.empty()` check I think, so you can't use that, and if you do you'll just get a subset of the actual cluster that is within limits; same with `GetDescendants`)

ajtowns replied:
see also the TODO item

> * Ability to inspect the would-be-constructed clusters (so they can be "fixed" if they violate policy rules, as opposed to just accepted/rejected, before applying - this is needed for reorgs which are not optional).

ajtowns replied:
:+1: I think it makes sense for TxGraph to take care of all the topology checks and feerate comparison decisions, which includes finding the best txs (block building), finding the worst txs (eviction) and finding good subsets when limits would otherwise be exceeded (reorgs, "rbf"/"carve out" cases maybe). So this doesn't feel like a design break at all to me, but rather pretty much exactly parallel to `GetWorstMainChunk()` -- "here's a bunch of txs that you want to remove".

instagibbs replied:
> so the only reasonable strategy I can see is to do:

The other reason to not follow this strategy, aside from reorgs, is "kindred eviction", where the caller is in charge of gathering sufficient information try to get back to an un-oversized state in the staging area, using non-oversized main stage information as a guiding tool.

ajtowns commented:
(conflicts with just merged #31519, Span to std::span)

ajtowns reviewed (commented):
ReACK 63b8f96e67f9ad649070a231532d48fb6c3573e4 . Needs a rebase anyway though.

The main/staging changes read better to me fwiw. The level stuff in the locators could be ungeneralised as well I suppose, but that doesn't seem likely to be much of an improvement.

ismaelsadeeq commented on src/txgraph.h:
```diff
@@ -0,0 +1,125 @@
+     *  arg does not exist. */
+    virtual FeeFrac GetIndividualFeerate(const Ref& arg) noexcept = 0;
+    /** Get pointers to all transactions in the connected component ("cluster") which arg is in.
+     *  The transactions will be returned in a topologically-valid order of acceptable quality.
```
"acceptable quality" is more than just a topologically valid order.
It refers to a linearization that may not be optimal but is the best we can achieve for that cluster within resource limitations.

ismaelsadeeq commented:
> That makes sense. Might be a good idea to update doc/developer-notes.md pointing out that use case? It currently says:

Done in #32100

instagibbs commented:
reviewed via `git range-diff master 72a97c0a07ea6e5a95ab37c8d95e1ea02cff8e92 1601906941fa559ebbee7898453fa77f4606ad38`

> Ref objects are now allowed to outlive TxGraph

Can this get surfaced in the fuzz coverage with something that would previously fail?

ajtowns commented:
ReACK 1601906941fa559ebbee7898453fa77f4606ad38

instagibbs commented:
reACK 1601906941fa559ebbee7898453fa77f4606ad38

> Dropping the TxGraphImpl destructor body then causes segfaults (rather than a nice assertion failure).

Right, I got myself turned around on this thinking it was harder to achieve. I'd suggest adding coverage here or in a quick follow-up.

sipa committed:
txgraph: Allow Refs to outlive the TxGraph (feature)

sipa reviewed (commented):
Changes:
* Did a big pass through the interface documentation in txgraph.h
* Moved the support for Refs outliving TxGraph to a separate commit, and added the test suggested in https://github.com/bitcoin/bitcoin/pull/31363#issuecomment-2742606528.
* Some smaller things listed below:

sipa committed:
txgraph: Add Get{Ancestors,Descendants}Union functions (feature)

Like GetAncestors and GetDescendants, but for the union of multiple inputs.

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -1942,6 +1957,13 @@ void TxGraphImpl::SanityCheck() const
     }
 }
 
+void TxGraphImpl::DoWork() noexcept
```
As it is right now, it's unused in the main PR.

I've been thinking when `DoWork` can be called in the mempool lifecycle and whether that would make a difference because everything is done lazily and we only apply staged changes when we want to get the current state. (specifically a scenario where after some lazy additions it cheap to just call `DoWork` and not wait until we want to get a state.)


Maybe after trimming due to a reorg, after multiple additions of transactions e.g loading new mempool.

ismaelsadeeq reviewed (commented):
re-ACK 41b4434fed169570ce0976c6e58db0d4a9614aaa  after https://github.com/bitcoin/bitcoin/pull/31363#pullrequestreview-2707582068

The new doc is better, and it's explicit that ref's can outlive their txgraph.

ajtowns commented on src/txgraph.h:
```diff
@@ -0,0 +1,212 @@
+    /** Get the feerate of the chunk which transaction arg is in the main graph. Returns the empty
+     *  FeePerWeight if arg does not exist in the main graph. The main graph must not be
+     *  oversized. */
+    virtual bool Exists(const Ref& arg, bool main_only = false) noexcept = 0;
```
Looks like comments for Exists/GetMainChunkFeerate/GetIndividualFeerate have been mixed up (in the "Add staging support" commit, I think)

ajtowns reviewed (changes requested):
range-diff review of 41b4434fed169570ce0976c6e58db0d4a9614aaa

Good to see it's now explicit which calls you can make on an oversized txgraph.

instagibbs commented:
reACK b2ea3656481b4196acaf6a1b5f3949a9ba4cf48f

nice doc improvements

`git range-diff master 1601906941fa559ebbee7898453fa77f4606ad38 b2ea3656481b4196acaf6a1b5f3949a9ba4cf48f`

ajtowns commented:
reACK b2ea3656481b4196acaf6a1b5f3949a9ba4cf48f

ismaelsadeeq reviewed (commented):
reACK  b2ea3656481b4196acaf6a1b5f3949a9ba4cf48f 🚀

glozow reviewed (commented):
ACK b2ea3656481

Reviewed the structure to convince myself this approach makes sense to support the features required, and reviewed the `SimTxGraph` + differential fuzzer in closer detail. I haven't read through all the `TxGraphImpl` functions but did some manual mutation testing.

glozow commented on src/cluster_linearize.h:
```diff
@@ -1336,6 +1336,38 @@ std::vector<ClusterIndex> MergeLinearizations(const DepGraph<SetType>& depgraph,
+        // in between forward.
+        while (place_before.Any()) {
+            // j cannot be 0 here; if it was, then there was necessarily nothing earlier which
+            // elem needs to be place before anymore, and place_before would be empty.
```
nit 0aa874a357865dd4768091f26dff238e66fb8d83
```suggestion
            // elem needs to be placed before anymore, and place_before would be empty.
```

glozow commented on src/test/fuzz/txgraph.cpp:
```diff
@@ -0,0 +1,732 @@
+                top_sim.AddDependency(par, chl);
+                real->AddDependency(*par, *chl);
+                break;
+            } else if (top_sim.removed.size() < 100 && command-- == 0) {
```
I was thinking that if all the transactions are descended from this tx, you'd add all of them to `to_remove` here. So if you started out with 99, you can end up with a lot more than 100 afterward. You wouldn't do more removals afterward, but you'd have more than 100.
